ppdet/modeling/backbones/bifpn.py

```diff
@@ -164,39 +183,80 @@ def __init__(self, num_chan, repeat=3, levels=5):
+                    epsilon=1e-04,
+                    param_attr=ParamAttr(
+                        initializer=Constant(1.0), regularizer=L2Decay(0.)),
+                    bias_attr=ParamAttr(regularizer=L2Decay(0.)),
+                    name='resample_bn_{}'.format(idx))
+            else:
+                feat = inputs[idx]
             feats.append(feat)
+        # Build additional input features that are not from backbone.
+        # P_7 layer we just use pool2d without conv layer & bn, for the same channel with P_6.
+        # https://github.com/google/automl/blob/master/efficientdet/keras/efficientdet_keras.py#L820
+        for idx in range(len(inputs), self.levels):
+            if feats[-1].shape[1] != self.num_chan:
+                feat = fluid.layers.conv2d(
+                    feats[-1],
+                    self.num_chan,
+                    filter_size=1,
+                    padding='SAME',
+                    param_attr=ParamAttr(initializer=Xavier()),
+                    bias_attr=ParamAttr(regularizer=L2Decay(0.)),
+                    name='resample_conv_{}'.format(idx))
+                feat = fluid.layers.batch_norm(
+                    feat,
+                    momentum=0.997,
+                    epsilon=1e-04,
+                    param_attr=ParamAttr(initializer=Constant(1.0), regularizer=L2Decay(0.)),
+                    bias_attr=ParamAttr(regularizer=L2Decay(0.)),
+                    name='resample_bn_{}'.format(idx))
+            feat = fluid.layers.pool2d(
+                feat,
+                pool_type='max',
+                pool_size=3,
+                pool_stride=2,
+                pool_padding='SAME',
+                name='resample_downsample_{}'.format(idx))
+            feats.append(feat)
+        # Handle the p4_2 and p5_2 with another 1x1 conv & bn layer
+        p4_2_p5_2 = []
+        for idx in range(1, len(inputs)):
+            feat = fluid.layers.conv2d(
+                inputs[idx],
+                self.num_chan,
+                filter_size=1,
+                padding='SAME',
+                param_attr=ParamAttr(initializer=Xavier()),
+                bias_attr=ParamAttr(regularizer=L2Decay(0.)),
+                name='resample2_conv_{}'.format(idx))
+            feat = fluid.layers.batch_norm(
+                feat,
+                momentum=0.997,
+                epsilon=1e-04,
+                param_attr=ParamAttr(initializer=Constant(1.0), regularizer=L2Decay(0.)),
+                bias_attr=ParamAttr(regularizer=L2Decay(0.)),
+                name='resample2_bn_{}'.format(idx))
+            p4_2_p5_2.append(feat)
 
-        biFPN = BiFPNCell(self.num_chan, self.levels)
+        biFPN = BiFPNCell(self.num_chan, self.levels, len(inputs))
         for r in range(self.repeat):
-            feats = biFPN(feats, 'bifpn_{}'.format(r))
+            if r == 0:
+                feats = biFPN(feats, cell_name='bifpn_{}'.format(r), is_first_time=True, p4_2_p5_2=p4_2_p5_2)
```

Comment:
just handle p4 and p5 when populate the `feats` list, so that there is no need for `is_first_time` and `p4_2_p5_2`?

Reply:
OK, and I would like to point out that it is different when the BIFPN has been used at the first time. Another 3rd party implementation could be referenced. https://github.com/bubbliiiing/efficientdet-pytorch/blob/7a89bf2b2f6b482b4d224f9c54a275e911234b71/nets/efficientdet.py#L153

Reply:
again, why can't be done when populate the initial `feats` list , before passing into the first bifpn cell?

Reply:
i have modified it in the third commit, have you seen?